07_web_endpoints/discord_bot.py

```diff
@@ -0,0 +1,230 @@
+#    On Modal's secret creation page, select 'Discord'. Copy your Discord application’s
+#    **Public Key** (in **General Information**) and paste the value of the public key
+#    as the value of the `DISCORD_PUBLIC_KEY` environment variable.
+#    Name this secret `weather-discord-secret`.
+
+# ## Register a Slash Command
+
+# Next, we’re going to register a command for our Discord app via an HTTP
+# endpoint.
+
+# Run the following command in your terminal, replacing the appropriate variable
+# inputs. `BOT_TOKEN` can be found by resetting your token in the application’s
+# **Bot** section, and `CLIENT_ID` is the **Application ID** available in
+# **General Information**.
+
+
+# ``` shell
+# BOT_TOKEN='replace_with_bot_token'
+# CLIENT_ID='replace_with_client_token'
+# curl -X POST \
+# -H 'Content-Type: application/json' \
+# -H "Authorization: Bot $BOT_TOKEN" \
+# -d '{
+#   "name":"get_weather",
+#   "description":"get weather",
+#   "options":[
+#     {
+#       "name":"city",
+#       "description":"The city for which you want to get the weather",
+#       "type":3,
+#       "required":true
+#     }
+#   ]
+# }' "https://discord.com/api/v10/applications/$CLIENT_ID/commands"
+# ```
+
+
+# This will register a Slash Command for your bot named `get_weather`, and has a
+# parameter called `city`. More information about the
+# command structure can be found in the Discord docs
+# [here](https://discord.com/developers/docs/interactions/application-commands).
+
+# ## Defining the asgi app with modal
+
+# We now create a `POST /get_weather` endpoint using [FastAPI](https://fastapi.tiangolo.com/) and Modal's
+# [@asgi_app](https://modal.com/docs/guide/webhooks#serving-asgi-and-wsgi-apps) decorator to handle
+# interactions with our Discord app (so that every time a user does a slash
+# command, we can respond to it).
+
+# Let's get the imports out of the way and define an [`App`](https://modal.com/docs/reference/modal.App)
+
+import json
+
+import modal
+
+app = modal.App("example-discord-weather-bot")
+
+# We define an [Image](https://modal.com/docs/guide/images) that has the [`python-weather`](https://github.com/null8626/python-weather) package and
+# the [FastAPI](https://fastapi.tiangolo.com/) package installed.
+
+image = modal.Image.debian_slim(python_version="3.11").pip_install(
+    "python-weather==2.0.7", "fastapi[standard]==0.115.4", "pynacl==1.5.0"
+)
+
+# We define a function that uses the `python_weather` library to get the weather of a city
+# Note that since Discord requires an interaction response within 3 seconds, we
+# use [`spawn`](https://modal.com/docs/reference/modal.Function#spawn) to kick off
+# `get_weather_for_city`as a background task from the asgi app while returning a `defer` message to
+# Discord within the time limit.
+
+
+@app.function(image=image)
+async def get_weather_forecast_for_city(
+    city: str, interaction_token, application_id
```

Comment:
This function is doing a bit too much -- figuring out the Discord URL to talk to, hitting a different external API, and then submitting the request. That makes it harder to write about and harder to effectively name.